datafusion/scheduler/src/task.rs

```diff
@@ -0,0 +1,225 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::{is_worker, spawn_local, spawn_local_fifo, Query};
+use futures::task::ArcWake;
+use log::{debug, trace};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Weak};
+use std::task::{Context, Poll};
+
+/// Spawns a query using the provided [`Spawner`]
+pub fn spawn_query(query: Arc<Query>) {
+    debug!("Spawning query: {:#?}", query);
+
+    let spawner = query.spawner();
+
+    for (pipeline_idx, query_pipeline) in query.pipelines().iter().enumerate() {
+        for partition in 0..query_pipeline.pipeline.output_partitions() {
+            spawner.spawn(Task {
```

Comment:
To check my understanding:

1. This code creates as many `Task`s  as there are "units of of work" -- aka (`pipeline`, `partition`) pairs
2. There is a big free-for-all race where each Task is asked "do you have any output yet"
3. If the answer to "yes" the scheduler calls "Pipeline::push` immediately with the output batch

Reply:
Pretty much 😃